test/e2e/all_in_one_test.go

```diff
@@ -336,52 +339,60 @@ func TestDeployAllInOneDBLESS(t *testing.T) {
 	t.Log("running ingress tests to verify all-in-one deployed ingress controller and proxy are functional")
 	deployIngress(ctx, t, env)
 	verifyIngress(ctx, t, env)
+	ensureAllProxyReplicasAreConfigured(ctx, t, env, deployments.ProxyNN)
 
-	const replicasCount = 3
-	gatewayDeployment := deployments.GetProxy(ctx, t, env)
-	gatewayDeployment.Spec.Replicas = lo.ToPtr(int32(replicasCount))
-	_, err := env.Cluster().Client().AppsV1().Deployments(gatewayDeployment.Namespace).Update(ctx, gatewayDeployment, metav1.UpdateOptions{})
-	require.NoError(t, err)
+	expectedControllerReplicas := *(deployments.GetController(ctx, t, env).Spec.Replicas)
+	t.Logf("store expected controller replicas count = %d", expectedControllerReplicas)
 
-	var podList *corev1.PodList
+	t.Log("scale proxy to 0 replicas")
+	scaleDeployment(ctx, t, env, deployments.ProxyNN, 0)
 
-	t.Log("waiting all the dataplane instances to be ready")
-	require.Eventually(t, func() bool {
-		appLabel := gatewayDeployment.Labels["app"]
-		forDeployment := metav1.ListOptions{
-			LabelSelector: fmt.Sprintf("app=%s", appLabel),
-		}
-		podList, err = env.Cluster().Client().CoreV1().Pods(gatewayDeployment.Namespace).List(ctx, forDeployment)
-		require.NoError(t, err)
+	t.Log("wait for 10 seconds to let controller reconcile")
+	<-time.After(10 * time.Second)
 
-		readyReplicasCount := lo.CountBy(podList.Items, func(pod corev1.Pod) bool {
-			return lo.ContainsBy(pod.Status.Conditions, func(cond corev1.PodCondition) bool {
-				return cond.Type == corev1.PodReady && cond.Status == corev1.ConditionTrue
-			})
-		})
-		return readyReplicasCount == replicasCount
-	}, time.Minute, time.Second)
+	t.Log("ensure that controller pods didn't crash after scaling proxy to 0")
+	ensureNoneOfDeploymentPodsHasCrashed(ctx, t, env, deployments.ControllerNN)
+
+	t.Log("scale proxy to 3 replicas and wait for all instances to be ready")
+	scaleDeployment(ctx, t, env, deployments.ProxyNN, 3)
+	ensureAllProxyReplicasAreConfigured(ctx, t, env, deployments.ProxyNN)
+}
 
-	t.Log("confirming that all dataplanes got the config")
+func ensureAllProxyReplicasAreConfigured(ctx context.Context, t *testing.T, env environments.Environment, proxyDeploymentNN types.NamespacedName) {
+	forDeployment := metav1.ListOptions{
+		LabelSelector: fmt.Sprintf("app=%s", proxyDeploymentNN.Name),
+	}
+	podList, err := env.Cluster().Client().CoreV1().Pods(proxyDeploymentNN.Namespace).List(ctx, forDeployment)
```

Comment:
I think we could use `getPodByLabels` to do this.

Reply:
Yeah, good point, let's reuse the goodies we already got in the tool belt :) d66c33d